pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py

```diff
@@ -0,0 +1,448 @@
+    Tester of multiple stage type case
+
+    Attributes:
+        dut (Dut): Object of the Device under test
+        test_menu (List[UnittestMenuCase]): The list of the cases
+        retry_count (int): number of retries for a single case when it is failed to start
+    """
+
+    def run_all_multi_stage_cases(
+        self,
+        reset: bool = False,
+        timeout: int = DEFAULT_TIMEOUT,
+        retry_count: int = DEFAULT_RETRY_COUNT,
+    ) -> None:
+        """
+        Run all multi_stage cases
+
+        Args:
+            reset: whether do a hardware reset before running the case
+            timeout: timeout in second
+            retry_count (int): number of retries for a single case when it is failed to start
+        """
+        for case in self.test_menu:
+            self.run_multi_stage_case(case, reset, timeout=timeout, retry_count=retry_count)
+
+    def run_multi_stage_case(
+        self,
+        case: UnittestMenuCase,
+        reset: bool = False,
+        timeout: int = DEFAULT_TIMEOUT,
+        retry_count: int = DEFAULT_RETRY_COUNT,
+    ) -> None:
+        """
+        Run a specific multi_stage case
+
+        Notes:
+            Will skip if the case type is not multi_stage
+
+        Args:
+            case: the specific case that parsed in test menu
+            reset: whether do a hardware reset before running the case
+            timeout: timeout in second
+            retry_count(int): number of retries for a single case when it is failed to start
+        """
+        if case.type == 'multi_stage':
+            additional_kwargs = {'case_start_time': time.perf_counter()}
+            if reset:
+                self.dut.serial.hard_reset()
+            for sub_case in case.subcases:
+                self.dut.expect(self.ready_pattern_list, timeout=timeout)
+                # Retry if write not success
+                for retry in range(retry_count):
+                    self.dut.write(str(case.index))
+                    try:
+                        self.dut.expect_exact(case.name, timeout=1)
+                        break
+                    except TIMEOUT as e:
+                        if retry >= retry_count - 1:
+                            raise e
+                self.dut.write(str(sub_case['index']))
+
+            self.dut.expect_unity_test_output(timeout=timeout, additional_kwargs=additional_kwargs)
+
+
+class MultiDevResource:
+    """
+    Resources of multi_dev dut
+
+    Attributes:
+        dut (Dut): Object of the Device under test
+        sem (Semaphore): Semaphore of monitoring whether the case finished
+        recv_sig (List[str]): The list of received signals from other dut
+        thread (Thread): The thread of monitoring the signals
+    """
+
+    def __init__(self, dut: Dut) -> None:
+        self.dut = dut
+        self.sem = Semaphore()
+        self.recv_sig: List[str] = []
+        self.thread: Thread = None  # type: ignore
+
+
+class MultiDevCaseTester(BaseTester):
```

Comment:
could you help create a GH issue "replace threads in `MultiDevCaseTester`" later after your PR got merged?

Reply:
>  after your PR got merged

anyway... I asked so because if this PR is not merged, the issue itself is meaningless.